eclass/rocm.eclass

```diff
@@ -0,0 +1,278 @@
+	done
+	ROCM_REQUIRED_USE+=" ) "
+
+	local flags=( "${ALL_AMDGPU_TARGETS[@]/#/amdgpu_targets_}" )
+	local optflags=${flags[@]/%/(-)?}
+	ROCM_USEDEP=${optflags// /,}
+}
+_rocm_set_globals
+unset -f _rocm_set_globals
+
+
+# @FUNCTION: get_amdgpu_flags
+# @USAGE: get_amdgpu_flags
+# @DESCRIPTION:
+# Convert specified use flag of amdgpu_targets to compilation flags.
+# Append default target feature to GPU arch. See
+# https://llvm.org/docs/AMDGPUUsage.html#target-features
+get_amdgpu_flags() {
+	local AMDGPU_TARGET_FLAGS
+	for gpu_target in "${ALL_AMDGPU_TARGETS[@]}"; do
+		local target_feature=
+		if use amdgpu_targets_${gpu_target}; then
+			case ${gpu_target} in
+				gfx906|gfx908)
+					target_feature=:xnack-
+					;;
+				gfx90a)
+					target_feature=:xnack+
+					;;
+				*)
+					;;
+			esac
+			AMDGPU_TARGET_FLAGS+="${gpu_target}${target_feature};"
+		fi
+	done
+	echo ${AMDGPU_TARGET_FLAGS}
+}
+
+# @FUNCTION: check_rw_permission
+# @USAGE: check_rw_permission <file>
+# @DESCRIPTION:
+# check read and write permissions on specific files.
+# allow using wildcard, for example check_rw_permission /dev/dri/render*
+check_rw_permission() {
+	[[ -r $1 ]] && [[ -w $1 ]] || die \
+		"Portage do not have read or write permissions on $1! \n Make sure both are in render group and check the permissions."
+}
+
+# == phase functions ==
+
+# @FUNCTION: rocm_src_configure
+# @DESCRIPTION:
+# configure rocm packages, and setting common cmake arguments
+rocm_src_configure() {
+	# allow acces to hardware
+	addpredict /dev/kfd
+	addpredict /dev/dri/
+
+	mycmakeargs+=(
+		-DCMAKE_INSTALL_PREFIX="${EPREFIX}/usr"
+		-DAMDGPU_TARGETS="$(get_amdgpu_flags)"
+		-DCMAKE_SKIP_RPATH=TRUE
+	)
+
+	CXX="hipcc" cmake_src_configure
+}
+
+# @FUNCTION: rocm_src_test
+# @DESCRIPTION:
+# Test whether valid GPU device is present. If so, find how to, and execute test.
+# ROCm packages can have to test mechanism:
+# 1. cmake_src_test. MAKEOPTS="-j1" ensures only one test on GPU at a time;
+# 2. one single gtest binary called "${PN,,}"-test;
+# 3. Some package like rocFFT have alternative test like rocfft-selftest;
+# 4. Custome testing binaries like dev-libs/rccl. Use ${ROCM_TESTS} to specify.
+rocm_src_test() {
+	# grant and check permissions on /dev/kfd and /dev/dri/render*
+	for device in /dev/kfd /dev/dri/render*; do
+		addwrite ${device}
+		check_rw_permission ${device}
+	done
+
+	: ${LD_LIBRARY_PATH:="${BUILD_DIR}/clients:${BUILD_DIR}/src:${BUILD_DIR}/library:${BUILD_DIR}/library/src:${BUILD_DIR}/library/src/device"}
+	export LD_LIBRARY_PATH
+	if grep -q 'build test:' "${BUILD_DIR}"/build.ninja; then
+		MAKEOPTS="-j1" cmake_src_test
+	elif [[ -d ${BUILD_DIR}/clients/staging ]]; then
+		cd "${BUILD_DIR}/clients/staging" || die "Test directory not found!"
+		for test_program in "${PN,,}-"*test; do
+			if [[ -x ${test_program} ]]; then
+					edob ./${test_program}
+			else
+				die "The test program ${test_program} does not exist or cannot be excuted!"
+			fi
+		done
```

Comment:
This part is oddly specific.

Reply:
Oh, there are two cases ROCm packages test themselves: standard cmake test, or custom test binaries. So `src_test` have to know which one the package uses, and choose the corresponding method.

Both situation needs to specify `LD_LIBRARY_PATH` (because `SKIP_RPATH` is set to `On`).

So `rocm_src_test` collect the two kinds of specific code from several ebuilds into one eclass, with extra codes to judge the occasion.

I'm not satisfied with this part, too. Any suggestions?